pkg/distributors/contracts/DistributionScheduler.sol

```diff
@@ -0,0 +1,142 @@
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+
+import "./interfaces/IMultiDistributor.sol";
+
+// solhint-disable not-rely-on-time
+
+/**
+ * Scheduler for MultiDistributor contract
+ */
+contract DistributionScheduler {
+    using SafeERC20 for IERC20;
+
+    IMultiDistributor private immutable _multiDistributor;
+
+    constructor(IMultiDistributor multiDistributor) {
+        _multiDistributor = multiDistributor;
+    }
+
+    enum DistributionStatus { UNINITIALIZED, PENDING, STARTED }
+
+    struct ScheduledDistribution {
+        bytes32 distributionId;
+        IERC20 stakingToken;
+        IERC20 distributionToken;
+        uint256 startTime;
+        address owner;
+        uint256 amount;
+        DistributionStatus status;
+    }
+
+    event DistributionScheduled(
+        bytes32 scheduleId,
+        address indexed owner,
+        IERC20 indexed stakingToken,
+        IERC20 indexed distributionToken,
+        uint256 startTime,
+        uint256 amount
+    );
+    event DistributionStarted(
+        bytes32 scheduleId,
+        address indexed owner,
+        IERC20 indexed stakingToken,
+        IERC20 indexed distributionToken,
+        uint256 startTime,
+        uint256 amount
+    );
+
+    mapping(bytes32 => ScheduledDistribution) private _scheduledDistributions;
+
+    function getScheduledDistributionInfo(bytes32 scheduleId)
+        external
+        view
+        returns (ScheduledDistribution memory reward)
+    {
+        return _scheduledDistributions[scheduleId];
+    }
+
+    function startDistributions(bytes32[] calldata scheduleIds) external {
+        for (uint256 i; i < scheduleIds.length; i++) {
+            bytes32 scheduleId = scheduleIds[i];
+            ScheduledDistribution memory scheduledDistribution = _scheduledDistributions[scheduleId];
+
+            require(scheduledDistribution.status == DistributionStatus.PENDING, "Distribution cannot be started");
+            require(scheduledDistribution.startTime <= block.timestamp, "Distribution start time is in the future");
+
+            _scheduledDistributions[scheduleId].status = DistributionStatus.STARTED;
+
+            uint256 allowance = scheduledDistribution.distributionToken.allowance(
+                address(this),
+                address(_multiDistributor)
+            );
+            if (allowance < scheduledDistribution.amount) {
+                scheduledDistribution.distributionToken.approve(address(_multiDistributor), type(uint256).max);
+            }
```

Comment:
I'd just do `approve(amount)`, as we do in the relayer, which means EIP 2929 savings will kick in (also resulting in simpler code).